File: src/main/java/com/dochiri/hexagonal/application/user/service/query/UserFinder.java
package com.dochiri.hexagonal.application.user.service.query;

import com.dochiri.hexagonal.application.user.port.out.LoadUserPort;
import com.dochiri.hexagonal.domain.user.User;
import com.dochiri.hexagonal.domain.user.exception.UserNotFoundException;
import com.dochiri.hexagonal.domain.user.vo.Email;
import com.dochiri.hexagonal.domain.user.vo.UserId;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
@RequiredArgsConstructor
public class UserFinder {

    private final LoadUserPort loadUserPort;

    @Transactional(readOnly = true)
    public User findById(UserId userId) {
        return loadUserPort.loadById(userId)
                .orElseThrow(() -> new UserNotFoundException("사용자를 찾을 수 없습니다: " + userId.value()));
    }

    @Transactional(readOnly = true)
    public User findByEmail(Email email) {
        return loadUserPort.loadByEmail(email)
                .orElseThrow(() -> new UserNotFoundException("사용자를 찾을 수 없습니다: " + email.value()));

    }

}

File: src/main/kotlin/com/dochiri/hexagonal/application/user/service/query/UserFinder.kt
package com.dochiri.hexagonal.application.user.service.query

import com.dochiri.hexagonal.domain.user.User
import com.dochiri.hexagonal.application.user.port.out.LoadUserPort
import com.dochiri.hexagonal.domain.user.exception.UserNotFoundException
import com.dochiri.hexagonal.domain.user.vo.Email
import com.dochiri.hexagonal.domain.user.vo.UserId
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
class UserFinder(
    private val loadUserPort: LoadUserPort
) {

    @Transactional(readOnly = true)
    fun findById(userId: UserId): User =
        loadUserPort.findById(userId) ?: throw UserNotFoundException.byId(userId)

    @Transactional(readOnly = true)
    fun findByEmail(email: Email): User =
        loadUserPort.findByEmail(email) ?: throw UserNotFoundException.byEmail(email)
}
